tests/rest/client/test_room_access_rules.py

```diff
@@ -670,6 +720,126 @@ def test_revoke_3pid_invite_direct(self):
+        requester = create_requester(self.user_id)
+        allowed_requester = create_requester("@user:allowed_domain")
+        forbidden_requester = create_requester("@user:forbidden_domain")
+
+        # Create a join event for a forbidden user
+        forbidden_join_event, _ = self.get_success(
+            event_creator.create_event(
+                forbidden_requester,
+                {
+                    "type": EventTypes.Member,
+                    "room_id": self.restricted_room,
+                    "sender": forbidden_requester.user.to_string(),
+                    "content": {"membership": Membership.JOIN},
+                    "state_key": forbidden_requester.user.to_string(),
+                },
+            )
+        )
+
+        # Create a join event for a not forbidden user
+        allowed_join_event, _ = self.get_success(
+            event_creator.create_event(
+                allowed_requester,
+                {
+                    "type": EventTypes.Member,
+                    "room_id": self.restricted_room,
+                    "sender": allowed_requester.user.to_string(),
+                    "content": {"membership": Membership.JOIN},
+                    "state_key": allowed_requester.user.to_string(),
+                },
+            )
+        )
+
+        # Create a state event defining a restricted room
+        restricted_access_rules_event = self.helper.get_state(
+            self.restricted_room, "im.vector.room.access_rules", self.tok
+        )
+
+        # Create a dict of fake state events for the room
+        room_state_events = {
+            (
+                EventTypes.Member,
+                requester.user.to_string(),
+            ): restricted_access_rules_event
+        }
+
+        # Assert a join event from a forbidden user to a restricted room is rejected
+        can_join = self.get_success(
+            access_rules.check_event_allowed(forbidden_join_event, room_state_events)
+        )
+        self.assertFalse(can_join)
+
+        # But a join event from an non-forbidden user to a restricted room is allowed
+        can_join = self.get_success(
+            access_rules.check_event_allowed(allowed_join_event, room_state_events)
+        )
+        self.assertTrue(can_join)
+
+        # Test that forbidden users can join unrestricted rooms if they have an invite
+
+        # Set up a fake state map for an unrestricted room
+        unrestricted_access_rules_event = self.helper.get_state(
+            self.unrestricted_room, "im.vector.room.access_rules", self.tok
+        )
+        room_state_events = {
+            (
+                EventTypes.Member,
+                requester.user.to_string(),
+            ): unrestricted_access_rules_event
+        }
+
+        # A forbidden user should not be able to join an unrestricted room
+        can_join = self.get_success(
+            access_rules.check_event_allowed(forbidden_join_event, room_state_events)
```

Comment:
The join event isn't for the right room here, though we don't necessarily care that much because `check_event_allowed` doesn't check on the room ID.